homeassistant/components/solax/sensor.py

```diff
@@ -0,0 +1,213 @@
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.exceptions import PlatformNotReady
+from homeassistant.helpers.event import async_track_time_interval
+
+_LOGGER = logging.getLogger(__name__)
+
+# key: name of sensor
+# value.0: index
+# value.1: unit (String) or None
+# from https://github.com/GitHobi/solax/wiki/direct-data-retrieval
+INVERTER_SENSORS = {
+    'PV1 Current':                (0, 'A'),
+    'PV2 Current':                (1, 'A'),
+    'PV1 Voltage':                (2, 'V'),
+    'PV2 Voltage':                (3, 'V'),
+
+    'Output Current':             (4, 'A'),
+    'Network Voltage':            (5, 'V'),
+    'Power Now':                  (6, 'W'),
+
+    'Inverter Temperature':       (7, TEMP_CELSIUS),
+    'Today\'s Energy':            (8, 'kWh'),
+    'Total Energy':               (9, 'kWh'),
+    'Exported Power':             (10, 'W'),
+    'PV1 Power':                  (11, 'W'),
+    'PV2 Power':                  (12, 'W'),
+
+    'Battery Voltage':            (13, 'V'),
+    'Battery Current':            (14, 'A'),
+    'Battery Power':              (15, 'W'),
+    'Battery Temperature':        (16, TEMP_CELSIUS),
+    'Battery Remaining Capacity': (17, '%'),
+
+    'Battery Energy':             (19, 'kWh'),
+
+    'Grid Frequency':             (50, 'Hz'),
+    'EPS Voltage':                (53, 'V'),
+    'EPS Current':                (54, 'A'),
+    'EPS Power':                  (55, 'W'),
+    'EPS Frequency':              (56, 'Hz'),
+}
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_IP_ADDRESS): cv.string,
+})
+
+SCAN_INTERVAL = timedelta(seconds=30)
+REQUEST_TIMEOUT = 5
+
+REAL_TIME_DATA_ENDPOINT = 'http://{ip_address}/api/realTimeData.htm'
+
+DATA_SCHEMA = vol.Schema(
+    vol.All([vol.Coerce(float)], vol.Length(min=68, max=68))
+)
+
+REAL_TIME_DATA_SCHEMA = vol.Schema({
+    vol.Required('method'): cv.string,
+    vol.Required('version'): cv.string,
+    vol.Required('type'): cv.string,
+    vol.Required('SN'): cv.string,
+    vol.Required('Data'): DATA_SCHEMA,
+    vol.Required('Status'): cv.positive_int,
+}, extra=vol.REMOVE_EXTRA)
+
+
+class SolaxRequestError(Exception):
+    """Error to indicate a Solax API request has failed."""
+
+    pass
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Platform setup."""
+    endpoint = RealTimeDataEndpoint(hass, config.get(CONF_IP_ADDRESS))
+    hass.async_add_job(endpoint.async_refresh)
+    async_track_time_interval(hass, endpoint.async_refresh, SCAN_INTERVAL)
+    devices = []
+    for sensor in INVERTER_SENSORS:
+        devices.append(Inverter(sensor))
+    endpoint.sensors = devices
+    async_add_entities(devices)
+
+
+async def async_solax_real_time_request(hass, schema, ip_address, retry,
```

Comment:
All communication with the device needs to be moved to a standalone python library hosted on pypi and not live within the Home Assistant source code. See: https://developers.home-assistant.io/docs/en/development_checklist.html for more details.

Reply:
Done!